### PR TITLE
chg: Cancel active sealing op when a new chain segment is received (MAT-1116)

### PR DESCRIPTION
**Pull request by atvanguard**

### CODE REVIEW

File: consensus/bor/bor.go [hunk 1]
```diff
@@ -774,23 +783,62 @@ func (c *Bor) Seal(chain consensus.ChainReader, block *types.Block, results chan
...
 	}()
-
 	return nil
 }
 
+// WaitForSealingOp blocks until delay elapses or stop signal is received
+func (c *Bor) WaitForSealingOp(stop <-chan struct{}, shouldSeal chan bool, number uint64, delay time.Duration) {
+	ctx, cancel := context.WithCancel(context.Background())
+	ctx = context.WithValue(ctx, blockNumberKey, number)
+	ctx = context.WithValue(ctx, listenerKey, shouldSeal)
```
Comment (atvanguard) [inline consensus/bor/bor.go:822]: I could change these right?

File: consensus/bor/bor.go [hunk 2]
```diff
@@ -774,23 +783,62 @@ func (c *Bor) Seal(chain consensus.ChainReader, block *types.Block, results chan
 
 	// Wait until sealing is terminated or delay timeout.
 	log.Trace("Waiting for slot to sign and propagate", "delay", common.PrettyDuration(delay))
+	shouldSeal := make(chan bool)
+	go c.WaitForSealingOp(stop, shouldSeal, number, delay)
 	go func() {
-		select {
-		case <-stop:
+		switch <-shouldSeal {
```
Comment (atvanguard) [inline consensus/bor/bor.go:789]: Should I `close(shouldSeal)` the channel immediately after this?

Comment (tokamak-git) [reply consensus/bor/bor.go:789]: if it is not going to be used anywhere else then yes, else defer would be a good idea.

File: consensus/bor/bor.go [hunk 3]
```diff
@@ -138,6 +138,13 @@ var (
 	errRecentlySigned = errors.New("recently signed")
 )
 
+type ActiveSealingOp struct {
+	ctx    context.Context
+	cancel context.CancelFunc
```
Comment (jdkanani) [inline consensus/bor/bor.go:143]: Why it needs `ctx` and `cancel`? Why not just `blockNumber` and `shouldSeal`?

File: consensus/bor/bor.go [hunk 4]
```diff
@@ -774,23 +783,62 @@ func (c *Bor) Seal(chain consensus.ChainReader, block *types.Block, results chan
 
 	// Wait until sealing is terminated or delay timeout.
 	log.Trace("Waiting for slot to sign and propagate", "delay", common.PrettyDuration(delay))
+	shouldSeal := make(chan bool)
+	go c.WaitForSealingOp(stop, shouldSeal, number, delay)
```
Comment (jdkanani) [inline consensus/bor/bor.go:787]: Shouldn't we cancel the previous `SealingOp` before we wait for new? Or is it done by the `stop` channel automatically?

Comment (atvanguard) [reply consensus/bor/bor.go:783]: Yes, the method that invokes this closes stop channel which causes the active one to stop before new sealing starts. https://github.com/maticnetwork/bor/blob/develop/miner/worker.go#L525

File: consensus/bor/bor.go (hunk 3 above)
Comment (atvanguard) [reply consensus/bor/bor.go]: Aah I don't need to use a cancellable object at all. Changed.

File: consensus/bor/bor.go (hunk 2 above)
Comment (atvanguard) [reply consensus/bor/bor.go:786]: Done.

File: consensus/bor/bor.go (hunk 1 above)
Comment (atvanguard) [reply consensus/bor/bor.go]: Actually I don't need to use a context at all. Removed it.

File: core/types.go [hunk 5]
```diff
@@ -49,3 +51,10 @@ type Processor interface {
 	// the processor (coinbase) and any included uncles.
 	Process(block *types.Block, statedb *state.StateDB, cfg vm.Config) (types.Receipts, []*types.Log, uint64, error)
 }
+
+// bor acts as a way to be able to type cast consensus.Engine;
+// since importing "github.com/maticnetwork/bor/consensus/bor" results in a cyclic dependency
+type bor interface {
```
Comment (tokamak-git) [inline core/types.go:57]: name could be `consense` instead of bor? interface being the verbe here.

Comment (atvanguard) [reply core/types.go]: Removed this altogether. Not the right place to be anyway.

File: consensus/bor/bor.go [hunk 6]
```diff
@@ -774,23 +781,64 @@ func (c *Bor) Seal(chain consensus.ChainReader, block *types.Block, results chan
...
+		shouldSeal <- false
+	case <-time.After(delay):
+		shouldSeal <- true
+	}
+}
+
+// CancelActiveSealingOp cancels in-flight sealing process
+func (c *Bor) CancelActiveSealingOp() {
+	if c.activeSealingOp != nil {
+		log.Info("Discarding active sealing operation", "number", c.activeSealingOp.number)
```
Comment (atvanguard) [inline consensus/bor/bor.go:837]: Should this be a debug statement instead?

Comment (jdkanani) [reply consensus/bor/bor.go:837]: Yes

Comment (atvanguard) [reply consensus/bor/bor.go]: Done.

**Pull request merged**